### PR TITLE
Prepend Silence to Audio in Cache Aware Training

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

NeMo cache inference path exposes `pad_and_drop_preencoded` which defaults to false, this aims to close the training-inference gap where the first chunk does not have pre-encoder left context. Exported models hardcode `pad_and_drop_preencoded=True` and do not have the ability to treat the first chunk differently which causes increased WER.

This PR closes the training-inference gap in a better way to reduce WER degradation for exported models
The following table compares two models that are identical on everything except this change, the WER numbers are from an internal dataset.

| inference path | before | after | delta |
| --- | ---: | ---: | ---: |
| offline | 0.08776 | 0.08848 | +0.07 pp |
| NeMo cache default (`pad_and_drop_preencoded=False`) | 0.08959 | 0.08909 | −0.05 pp |
| pad+drop streaming, **log zeros** (current `main`)| 0.12190 | **0.08844** | **−3.35 pp** |
| pad+drop streaming, **log silence** | 0.11350 | **0.08802** | **−2.55 pp** |

This change is completely backwards compatible, old models will work as they used to and new models can still use the old inference path, but new models will perform much better when exported to ONNX/TRT/CUDA Graphs

**Collection**: ASR
# Changelog

- Prepend silence as pre-encoder cache in training

# Usage

On by default and backwards compatible, no changes needed user-side

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Speech/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

@nithinraok @pzelasko @artbataev @ipmedenn

### CODE REVIEW

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 1]
```diff
@@ -61,6 +61,9 @@
 
 __all__ = ['ConformerEncoder', 'ConformerMultiLayerFeatureExtractor']
 
+# Precomputed log-mel feature value of digital silence.
+LOG_MEL_SILENCE = math.log(2**-24)
```
Comment (nithinraok) [inline nemo/collections/asr/modules/conformer_encoder.py:65]: Could we make this a configured feature-space value instead of a `ConformerEncoder` constant? The encoder input can use preprocessing or normalization that does not map silence to `log(2**-24)`. The affected `normalize: "NA"` configurations can explicitly set `pre_encode_pad_value: -16.635532333438686`. `CacheAwareStreamingAudioBuffer` should use the same configured value instead of always inserting `0.0`.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 2]
```diff
@@ -677,10 +680,22 @@ def forward_internal(
             if isinstance(self.pre_encode, nn.Linear):
                 audio_signal = self.pre_encode(audio_signal)
             else:
+                # Later chunks carry pre_encode_cache_size frames over from the previous chunk.
+                # The first one has nothing to carry over, so pad it with silence and let the drop
+                # below handle it like any other chunk. Same as pad_and_drop_preencoded=True.
+                if self.training and self.att_context_style == "chunked_limited":
```
Comment (nithinraok) [inline nemo/collections/asr/modules/conformer_encoder.py:686]: Could we make this behavior explicit and keep the existing training behavior by default? For example, add `pad_and_drop_preencoded: bool = False` to the encoder configuration and include it in this condition. Existing configurations would remain unchanged, while the affected export configurations could enable the new behavior.

**Review by nithinraok — CHANGES_REQUESTED** [review]
Thanks for addressing the first-chunk training/inference mismatch. The issue is valid, and the pad-and-drop approach makes sense for a fixed-shape exported graph.

Could you clarify whether the two `pad+drop streaming` rows were measured with an actual exported ONNX/TRT/CUDA Graph model, or with NeMo/PyTorch cache-aware streaming using `pad_and_drop_preencoded=True`?

I added two inline comments about making the new behavior explicit and using one configurable padding value across training and streaming.

Please also add a CPU regression test for this behavior.

**Comment by MahmoudAshraf97** [comment]
The numbers are deliberately measured using Nemo streaming iterator to avoid attributing the issue to compilation or export arifacts, although I reverified on TRT and they still hold to a certain numeric tolerance.
I will add the test soon.

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 1 above)
Comment (MahmoudAshraf97) [reply nemo/collections/asr/modules/conformer_encoder.py:65]: Two cases here:
1. During training, the only two things that can affect this value are `log_zero_guard_type` and `log_zero_guard_value`, both of them are preprocessor flags that the encoder knows nothing about, so during training we have to pass them using either a config value or precalculate them with preprocessor and pass that value, in both cases the encoder gets another argument and if we use the preprocessor to precalculate the value, the implementation now needs to exist in the init of each affected model class
2. During inference, old models will not have that flag, so they will still use zeros or another default, which brings us back to the same point of the padding value being a constant, and new models are indifferent to the padding value so they don't need a dynamic one.

The value invariance was tested in the table above where the new model was padded with zeros in mel space and still performed as expected, although the whole training distribution did not contain mel zeros so it's an OOD padding value.

The benefit from this PR comes mainly from the pad and drop during training, not from the padding value, this value was only chosen because all published streaming configs will default to it.

Since the scope of this change is cache-aware models, it's expected that these models should use NA normalization, otherwise there is already a much bigger train-inference mismatch resulting from chunk vs whole utterance normalization or even online normalization, which makes the constant mismatch negligible in comparison.
Also if we allow the users to set this, we must check it against the actual value that would've been produced by the preprocessor, the check needs the preprocessor's log_zero_guard_type/log_zero_guard_value inside the encoder, which is exactly the coupling the config value was introduced to avoid.

Comment (MahmoudAshraf97) [reply nemo/collections/asr/modules/conformer_encoder.py:65]: Updating `CacheAwareStreamingAudioBuffer` with the new value is another point and I can update it with a constant too if you agree that a constant is enough in this case, given that this also only helps old models as new ones will be invariant to it

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 2 above)
Comment (MahmoudAshraf97) [reply nemo/collections/asr/modules/conformer_encoder.py:686]: I think this behavior should not be optional, it's considered a bugfix not a feature (the same class as #13827), so the old behavior does not need to be preserved, it's also backward compatible, old models will run exactly as they do now, and new models will benefit from better export robustness
